kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -79,6 +83,354 @@ public static void validateSchema(StructType schema, boolean isColumnMappingEnab
+      Metadata metadata) {
+    checkArgument(
+        isColumnMappingModeEnabled(
+            ColumnMapping.getColumnMappingMode(metadata.getConfiguration())));
+    validateSchema(newSchema, true /*columnMappingEnabled*/);
+    validatePartitionColumns(newSchema, new ArrayList<>(expectedPartitionColumns));
+    validateSchemaEvolution(
+        currentSchema,
+        newSchema,
+        ColumnMapping.getColumnMappingMode(metadata.getConfiguration()),
+        metadata);
+  }
+
+  /* Validate if a given schema evolution is safe for a given column mapping mode*/
+  private static void validateSchemaEvolution(
+      StructType currentSchema,
+      StructType newSchema,
+      ColumnMappingMode columnMappingMode,
+      Metadata metadata) {
+    switch (columnMappingMode) {
+      case ID:
+      case NAME:
+        validateSchemaEvolutionById(currentSchema, newSchema, metadata);
+        return;
+      case NONE:
+        throw new UnsupportedOperationException(
+            "Schema evolution without column mapping is not supported");
+      default:
+        throw new UnsupportedOperationException(
+            "Unknown column mapping mode: " + columnMappingMode);
+    }
+  }
+
+  /* Validates a given schema evolution by using field ID
+  as the source of truth for identifying fields */
+  private static void validateSchemaEvolutionById(
+      StructType currentSchema, StructType newSchema, Metadata metadata) {
+    Map<Long, StructField> currentFieldsById = fieldsById(columnPathToField(currentSchema));
+    Map<Long, StructField> updatedFieldsById = fieldsById(columnPathToField(newSchema));
+    SchemaChanges schemaChanges = computeSchemaDiffById(currentFieldsById, updatedFieldsById);
+
+    // Validate physical name consistency on the updated fields
+    validatePhysicalNameConsistency(schemaChanges.updatedFields());
+
+    // If IcebergCompatV2 validate nested ID existence and consistency for relevant fields
+    if (TableConfig.ICEBERG_COMPAT_V2_ENABLED.fromMetadata(metadata)) {
+      validateNestedIds(schemaChanges, currentFieldsById, updatedFieldsById);
+    }
+
+    // Validate the schema against the diff to make sure there's no type or nullability changes
+    validateUpdatedSchemaCompatibility(schemaChanges);
+  }
+
+  /* Compute the SchemaChanges using field IDs */
+  private static SchemaChanges computeSchemaDiffById(
+      Map<Long, StructField> currentFieldIdToField, Map<Long, StructField> updatedFieldIdToField) {
+    SchemaChanges.Builder schemaDiff = SchemaChanges.builder();
+    for (Map.Entry<Long, StructField> fieldInUpdatedSchema : updatedFieldIdToField.entrySet()) {
+      StructField existingField = currentFieldIdToField.get(fieldInUpdatedSchema.getKey());
+      StructField updatedField = fieldInUpdatedSchema.getValue();
+      // New field added
+      if (existingField == null) {
+        schemaDiff.withAddedField(updatedField);
+      } else if (isUpdatedField(existingField, fieldInUpdatedSchema.getValue())) {
```

Comment:
can we tell if the field is moved from one struct to another struct without looking at the full path?

Reply:
Good call out, I think this is an important one to follow up on. Though I think the solution isn't to look at the full path, I think we'd need to keep track of field ID to parent field ID. We'd then compare if the target field ID of a move is the same as the parent ID of the field being moved. If it's different that means we're trying to move a field outside of a struct which means it's invalid and should fail.

I think the TLDR is we'll need more encapsulation for representing move operations and making sure those are safe.